### ISSUE TITLE
psutil inside docker container reports host stats

### ISSUE THREAD

**Issue by avaranovich** [giampaolo/psutil#1011]
If I use psutil inside a docker container, it reports the **host** machine stats, i.e. CPU and RAM.
I would expect it to grab a container-level statistic.

**Comment by giampaolo** [comment]
You can change `/proc` filesystem location like this:

```
import psutil
psutil.PROCFS_PATH = "/something_else"
```

See:
https://pythonhosted.org/psutil/#psutil.PROCFS_PATH
https://github.com/giampaolo/psutil/issues/558
Please note that most information on Linux is retrieved via /proc but not all of it (e.g. `swap_memory`, `net_if_addrs`, `net_if_stats`, `disk_partitions` and others) so the PROCFS_PATH trick does not cover all psutil API.

**Comment by giampaolo** [comment]
Update: `swap_memory()` now relies on /proc, see #1015.

**Comment by crazyglasses** [comment]
psutil.PROCFS_PATH = "/something_else"

hey, for my container, what would be the '/something_else' ?

**Comment by giampaolo** [comment]
I don't know. I don't have a Docker instance to check against. Try to google for "docker proc filesystem" to see where it is located.

**Comment by keitwb** [comment]
@crazyglasses You will have to mount the root mount namespace's `/proc` dir into your container with the `-v` flag to `docker run`. E.g. `docker run -v /proc:/host-proc ...` and then configure PROCFS_PATH to point to that.

**Comment by crazyglasses** [comment]
@keitwb  Basically what I want to do is run psutil inside a python script inside a container. A few comments that I have gotten is that by default the psutil is pointed towards the containers /proc and that I would get metrics of the container if I use it that way . What you are suggesting seems to be mounting the root inside the container, doesnt that mean I get the host stats only?

**Comment by keitwb** [comment]
@crazyglasses Sorry, misread your question.  For docker containers, there isn't a hard distinction between CPU/mem inside or outside the container like there would be on a VM -- all containers share the same kernel and there isn't a namespace for cpus or memory that causes the container to see those differently.  There are cgroups, which you can use to limit resource usage of the containers (e.g. certain cores or period/quota of CPU), but that would just show as the processes in the container using a more limited amount of resources, not change what psutil would report as far as I understand it.  I know `docker stats` will give you what it sounds like you want -- I think it relies on cgroup stats somewhere in the proc filesystem, but I'm not sure if psutil can pick up those or not.

**Comment by ZoeShaw101** [comment]
So `psutil` run in container collects **host stats** or **container stats** ?

**Comment by giampaolo** [comment]
Container.

**Comment by ZoeShaw101** [comment]
@giampaolo So how can I collect host stats ? thanks

**Comment by giampaolo** [comment]
I don’t know enough about containers but I guess you can mount hosts’s /proc fs so that it’s visibile from within the container. With that in place you can:

‘’’
import psutil
psutil.PROCFS_PATH = “/path/to/host/procfs”
‘’’

From then on most psutil’s API will return stats about the host. All of this is based on the assumption that host and container’s /proc filesystems are independent, of which I am not completely sure. If they are querying /proc for things such as system wide stats (CPU, memory...) is safe but querying processes is not.

**Comment by ZoeShaw101** [comment]
OK, thanks alot.

**Comment by jankatins** [comment]
> All of this is based on the assumption that host and container’s /proc filesystems are independent, of which I am not completely sure

It seems that this is not correct: This blog posts shows that /proc/meminfo shows the information from the host and not the container: https://ops.tips/blog/why-top-inside-container-wrong-memory/ You can get the memory via the proposed way in https://github.com/giampaolo/psutil/issues/490

**Comment by gb160** [comment]
> I don’t know enough about containers but I guess you can mount host’s /proc fs so that it’s visibile from within the container. With that in place you can:
> 
> ‘’’ import psutil psutil.PROCFS_PATH = “/path/to/host/procfs” ‘’’
> 
> From then on most psutil’s API will return stats about the host. All of this is based on the assumption that host and container’s /proc filesystems are independent, of which I am not completely sure. If they are then querying /proc for things such as system wide stats (CPU, memory, ...) is safe, but querying processes is not.

apologies for resurrecting this, but can I use the proc trick to use psutil.net_io_counters() and get the host machines bytes_sent/bytes_recievd ?

**Comment by giampaolo** [comment]
Yes, you should because psutil.net_io_counters() retrieves its data from /proc/net/dev.

**Comment by gb160** [comment]
> Yes, you should because psutil.net_io_counters() retrieves its data from /proc/net/dev.

Yeah I thought the same but it doesn't work, psutil.net_io_counters() only returns the data from the container. There is something weird going on that's above my paygrade.

The only way I could get it to work was to start the container using '--net=host', which gets psutil.net_io_counters() retrieving the host network info, but that isn't ideal.

**Comment by giampaolo** [comment]
Also see discussion at https://github.com/giampaolo/psutil/issues/2100.

**Comment by giampaolo** [comment]
Closing this as a duplicate of https://github.com/giampaolo/psutil/issues/2100, which has more info.
